### PR TITLE
[SPARK-33573][SHUFFLE][YARN] Shuffle server side metrics for Push-based shuffle

### PR DESCRIPTION
**Pull request by rmcyang**
### What changes were proposed in this pull request?
This is one of the patches for SPARK-33235: Push-based Shuffle Improvement Tasks.
Added a class `PushMergeMetrics`, to collect below metrics from shuffle server side for Push-based shuffle:
- no opportunity responses
- too late responses
- pushed bytes written
- deferred block bytes
- number of deferred blocks
- stale block push requests
- ignored block bytes
### Why are the changes needed?
This helps to understand the push based shuffle metrics from shuffle server side.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added a method `verifyMetrics` to verify those metrics in existing unit tests.

### CODE REVIEW

**Comment by rmcyang** [comment]
cc @zhouyejoe @otterc @mridulm @xkrogen. Please take a look, thanks.

**Comment by zhouyejoe** [comment]
There were some test failures in YarnShuffleServiceSuite from some of the UTs adde by me in the NodeManager work preserving restart PR. Reran the PR test job to see whether it is consistent failing. @rmcyang  Are you able to reproduce the UT failures locally?

**Comment by rmcyang** [comment]
@zhouyejoe Hm not really,  running those UTs locally looks to me.
```
minyang@minyang-mn3 spark % build/mvn -Dtest=none -Pyarn -Phadoop-2.10 -Dscala-2.12 -Phive -Phive-2.3 -DwildcardSuites=org.apache.spark.network.yarn.YarnShuffleServiceWithLevelDBBackendSuite test -pl 'resource-managers/yarn'
Using `mvn` from path: /Users/minyang/project/OSS-linkedin/spark/build/apache-maven-3.8.6/bin/mvn
...
[INFO] Skipping execution of surefire because it has already been run for this configuration
[INFO] 
[INFO] --- scalatest-maven-plugin:2.1.0:test (test) @ spark-yarn_2.12 ---
[INFO] ScalaTest report directory: /Users/minyang/project/OSS-linkedin/spark/resource-managers/yarn/target/surefire-reports
Discovery starting.
Discovery completed in 670 milliseconds.
Run starting. Expected test count is: 17
YarnShuffleServiceWithLevelDBBackendSuite:
- executor and merged shuffle state kept across NM restart
- removed applications should not be in registered executor file and merged shuffle file
- shuffle service should be robust to corrupt registered executor file
- get correct recovery path
- moving recovery file from NM local dir to recovery path
- service throws error if cannot start
- Consistency in AppPathInfo between in-memory hashmap and the DB
- Finalized merged shuffle are written into DB and cleaned up after application stopped
- Dangling finalized merged partition info in DB will be removed during restart
- Dangling application path or shuffle information in DB will be removed during restart
- Cleanup for former attempts local path info should be triggered in applicationRemoved
- recovery db should not be created if NM recovery is not enabled
- SPARK-31646: metrics should be registered into Node Manager's metrics system
- SPARK-34828: metrics should be registered with configured name
- create default merged shuffle file manager instance
- create remote block push resolver instance
- invalid class name of merge manager will use noop instance
Run completed in 2 seconds, 944 milliseconds.
Total number of tests run: 17
Suites: completed 2, aborted 0
Tests: succeeded 17, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  24.019 s
[INFO] Finished at: 2022-08-30T02:58:38-07:00
[INFO] ------------------------------------------------------------------------
[WARNING] The requested profile "hadoop-2.10" could not be activated because it does not exist.
[WARNING] The requested profile "hive" could not be activated because it does not exist.
[WARNING] The requested profile "hive-2.3" could not be activated because it does not exist.
```

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java
Comment (mridulm) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:305]: We should not be directly creating `RemoteBlockPushResolver` within service init - it is done via `newMergedShuffleFileManagerInstance`, so that it returns the right instance configured by admin (by default disabled).

We should be adding a `getMetrics` to `MergedShuffleFileManager` to retrieve the metrics we want to register.

Comment (rmcyang) [reply common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java]: Thanks, the UTs passed after the fix.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1955]: `pushedBytesWritten` -> `blockBytesWritten` ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1163]: There are a bunch of other places where we simply drop `deferredBufs` entirely (`abortIfNecessary`, `onData`, etc).
Given this is tracked as a counter, we should update there as well (since we have already `inc` on it).

Essentially, replace all `deferredBufs = null` with `freeDeferredBufs`

**Review by mridulm — COMMENTED** [review]
Mostly looks good, added a few comments.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1163]: Makes sense. Have checked all the places and made the replacement, it seems just applies to `abortIfNecessary` and `onData`.

Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1163]: Makes sense. Have checked all the places and made the replacement, looks like the remaining places that need it are only `abortIfNecessary` and `onData`.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:612]: li-2.3.0 has the corresponding implementation as below:
```
public void onComplete(String streamId) {
  if (isTooLate) {
    pushMergeMetrics.tooLateResponses.mark();
    // Throw an exception here so the block data is drained from channel and server
    // responds RpcFailure to the client.
    throw new RuntimeException(String.format("Block %s %s", msg.blockId,
      BlockPushException.TOO_LATE_MESSAGE_SUFFIX));
  }
  // For duplicate block that is received before the shuffle merge finalizes, the
  // server should respond success to the client.
}
```
I think the reason we are doing this only as part of `onComplete` is that we only mark tooLateResponse when all data from the stream has been received, and ignore those for the callbacks of `onData` and `onFailure`, correct? @otterc

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: > Additionally, do we want to track the number of blocks which are deferred ?

Not sure if this would be a bit redundant, given we have tracked the deferred blocks in bytes, @otterc thoughts?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:612]: That looks like an impl detail in li-2.3.0 branch - any reason not to update this sooner ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: There is a correlation between number of deferred blocks and amount of bytes deferred - but they are not necessarily the same, and also are influenced by the spark application configuration.

**Review by mridulm — COMMENTED** [review]
Mostly looks good, just a few comments.

+CC @otterc, @Ngone51 for review

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:612]: This is explained in the comments above that why we don't want throw exception and close the channel but rather respond with stream callback and only when the complete block is received, respond with tooLate

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1232]: If too late, mark `lateBlockPushes` ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1231]: QQ: Do we want to capture stale pushes ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: The reason to track deferred bytes is because we want to know how much memory can be used up by these blocks. What insight can we derive from tracking the number of these blocks?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:612]: Yes, but is there a reason to defer updating the metric to later ? It is clear at time of callback construction itself that it is late.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: As I mentioned above, they are related but not the same - particularly when applications do not have homogeneous configurations.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (zhouyejoe) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:440]: Nit: missing one space.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (zhouyejoe) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1443]: Looks like the expectedCachedBlocksBytes are all 0 in all the callers, do we need to add any UTs for testing the CachedBytes metrics?

Comment (rmcyang) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1445]: Sure, I enhanced the UTs `testDeferredBufsAreWrittenDuringOnData` and `testDeferredBufsAreWrittenDuringOnComplete` to cover the test for `expectedCachedBlocksBytes` since the UTs produced the cached block bytes. Thanks for pointing this out.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:612]: Any comments on this @otterc, @zhouyejoe ?
Essentially, we are not marking late block push in case of `onFailure` - any reason for it ?

IMO we should do it for both - and if yes, we can move it out of the callback entirely and do it immediately before the callback construction.
Pls let me know if I am missing something ! Thx

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Can we add `deferredBlocks` as well ? Which track the number of deferred blocks (in addition to the size you had added support for here) ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1231]: Any thoughts ? Thx

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1232]: Any thoughts ?

**Review by mridulm — COMMENTED** [review]
Just a couple of queries and comments.
Once addressed, should be ready to merge. 
Thanks for working on this @rmcyang

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Yes, we should mark `lateBlockPushes` here, thanks for the suggestion!

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:612]: Actually in our 2.3 version, we were incrementing this metric whenever, the server was responding to the client that the block is tooLate. In `onFailure` we don't respond to the client with `TooLate`. However, these metrics are the server's view so I think you are right. We should increment outside the callback.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Thanks for the suggestion. It seems that a stale block push can appear during indeterminate stage retries ([reference](https://github.com/apache/spark/blob/05cd5f97c3dea25dacdbdb319243cdab9667c774/common/network-common/src/main/java/org/apache/spark/network/server/BlockPushNonFatalFailure.java#L132)). Not sure how necessary capturing this would be, it can be captured similar to capturing `lateBlockPushes` @mridulm .

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:612]: Can you make this change @rmcyang  ? Thx

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: I would look at these metrics as a means for identifying the effectiveness of push based shuffle, what is impacting the inefficiencies (how many stale blocks, how many late blocks, how that distribution is changing with time), etc.
These will correlate in some way to overall impact on when push based shuffle was used, when normal shuffle was used, and what the network and task impact of those are.

Do let me know if you have thoughts !

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Thanks for the comments. Fixed.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Thanks for the suggestion. Added `deferredBlocks`, PTAL.
WIP on adding UT.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Make sense. Added `staleBlockPushes`, PTAL.
WIP to add UT.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:588]: Even if `partitionInfoBeforeCheck.mapTracker.contains(msg.mapIndex)`, it would be considered late.
Move `pushMergeMetrics.lateBlockPushes.mark();` to the `else` block below (before `new StreamCallbackWithID`) ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1263]: This is when the deferred block is getting written out.
Instead, we should be incrementing it when the deferred block is getting added.
Move it to where we increment `deferredBlockBytes` below ?
This will keep the metrics in sync - else we can have a large `deferredBlockBytes`, but very low `deferredBlockBytes`.

An example of this is in the test case [here](https://github.com/apache/spark/pull/37638/files#diff-955b4eb8f495555d5e188ed93a4acf5393e1d9c7855dffddedb5e9201eebdef2R193) - the deferred blocks should be `1`m but is `0`

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (mridulm) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:193]: I have commented about this earlier in `RemoteBlockPushResolver` - and this will change when it is addressed.

**Review by mridulm — COMMENTED** [review]
Couple of comments.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (mridulm) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:277]: We should be capturing this in metrics - essentially streams which were ignored due to a speculative task writing concurrently.
We can add it as a follow up though.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (mridulm) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:254]: This will change with [the comment ](https://github.com/apache/spark/pull/37638#discussion_r1057874346) above ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1971]: We should be documenting these in `monitoring.md`

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Sure. In addition to the proposal here, to keep the `deferredBlockBytes` and `deferredBlocks ` fully in sync, I also went ahead and decrease the `deferredBlocks` whenever there is a `deferredBlockBytes.dec()` and accommodate the UTs, please take a look.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1969]: Sounds good, I added a few lines describing them under `shuffleService` section, PTAL.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (rmcyang) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:277]: > We should be capturing this in metrics

Not sure if I'm following here, could you please elaborate? Thanks @mridulm !

Comment (mridulm) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:277]: The specific testcase here is, we have duplicate blocks being sent (via speculative execution for example), and so we want to ignore data from the "newer" push. For example, here we stream2 would be ignored. The metrics are not capturing it - and so we see the same result in metrics as what we see for a successful push (what we would see for `testBasicBlockMerge`).

**Comment by mridulm** [comment]
Can you fix the linter error @thejdeep ?
` src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:[1281] (sizes) LineLength: Line is longer than 100 characters (found 104).`

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (rmcyang) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:277]: Thanks @mridulm. So you are suggesting to add a new metric to capture size/byte of pushed blocks that got ignored, right? If so, I'm trying to figure out what insights we could get if we add number/size of ignored blocks, thoughts? ++ @otterc

**Review by zhouyejoe — COMMENTED** [review]
Mostly LGTM other than some minor comments.

File: docs/monitoring.md
Comment (rmcyang) [reply docs/monitoring.md]: Sure, below is the preview after the change, PTAL.
Screen Shot 2023-01-10 at 12 15 47 PM

File: docs/monitoring.md
Comment (zhouyejoe) [inline docs/monitoring.md:1428]: These metrics are not related to client side configurations, as if push is not enabled in client side, these metrics will still be shown up, but all 0.  Once server side configuration is set correctly to MergedShuffleFileManager, these metrics will be shown up.

```suggestion
- **notes:** The metrics below are specific to push-based shuffle and only emitted when 
  `spark.shuffle.push.server.mergedShuffleFileManagerImpl` is configured as 
  "org.apache.spark.network.shuffle.RemoteBlockPushResolver".
```
Please refer to other notes format in this file. You do need to follow the format there.

File: docs/monitoring.md
Comment (zhouyejoe) [inline docs/monitoring.md:1436]: Remove the indent. These metrics should be within the same level as other metrics in shuffle service.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java
Comment (rmcyang) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java]: Sounds good, added a new metric `ignoredBlockBytes` to tracks the size of the blocks that are ignored. @mridulm @otterc

File: docs/monitoring.md
Comment (rmcyang) [reply docs/monitoring.md]: I removed the client side config in the description. As for the format for note, I'm following the format at L1408, L1399, L1392, and L1384. This is the preview after the change:
Screen Shot 2023-01-10 at 4 23 19 PM
Let me know if you have further input, thanks @zhouyejoe !

File: docs/monitoring.md
Comment (rmcyang) [reply docs/monitoring.md]: Fixed.

File: docs/monitoring.md
Comment (mridulm) [inline docs/monitoring.md:1437]: nit: 
Can we reorder this so that we go from the common metrics to the more corner cases ?
For example, something like:

* blockBytesWritten
* blockAppendCollisions
* lateBlockPushes
* staleBlockPushes
* deferredBlocks
* deferredBlockBytes
* ignoredBlockBytes

(If this is not the right order, please feel free to reorder them)

**Review by mridulm — APPROVED** [review]
Looks good to me - had one nit comment.

File: docs/monitoring.md
Comment (rmcyang) [reply docs/monitoring.md:1435]: Thanks for the suggestion, I further put the `staleBlockPushes` to the middle of `deferredBlockBytes ` and `ignoredBlockBytes`, as a stale block push appears in the case of indeterminate stage retries, which I feel is less common to deferred block. Let me know if that's not right. Thanks.

Comment (rmcyang) [reply docs/monitoring.md:1435]: Also adjust a bit after @zhouyejoe's suggestion, preview of the corresponding section:
Screen Shot 2023-01-11 at 2 50 26 PM

Comment (zhouyejoe) [reply docs/monitoring.md:1435]: LGTM.

**Review by zhouyejoe — APPROVED** [review]

File: docs/monitoring.md
Comment (mridulm) [reply docs/monitoring.md:1435]: Can we add a new line separation from the earlier list (before the `note: the metrics ...`) ?
All metrics following the note are related to push based shuffle - would be good to visually group them together.

Comment (rmcyang) [reply docs/monitoring.md:1437]: Sounds good. Current preview:
Screen Shot 2023-01-11 at 6 31 34 PM

**Review by mridulm — APPROVED** [review]
Looks good to me.
Will merge once tests complete

**Comment by mridulm** [comment]
Merged to master.
Thanks for working on this @rmcyang !
Thanks for the reviews @zhouyejoe, @otterc :-)

**Pull request closed without merging**
